Change ReactChildrenMutationWarningHook to Object.freeze children

keyanzhang opened the pull request:
Alternative of #7410 that fixes #7406. This should also fix #7424 and #7426 (the false-positive warning).
Saving and comparing the `shadowChildren` is pretty expensive -- we can `Object.freeze` the JSX children instead.

Benchmark:
```
Comparing master.txt (control) vs freeze.txt (test)
Significant differences marked by ***
% change from control to test, with 99% CIs:

* factory_ms_jsc_jit
    % change:  +0.38% [ -6.59%,  +7.36%]
    means: 27.6427 (control), 27.7558 (test)
* factory_ms_jsc_nojit
    % change: -10.22% [-15.03%,  -5.41%]  ***
    means: 26.3667 (control), 23.6813 (test)
* factory_ms_node
    % change:  -9.60% [-13.73%,  -5.46%]  ***
    means: 98.522 (control), 89.092 (test)
* ssr_pe_cold_ms_jsc_jit
    % change:  -8.47% [-11.57%,  -5.37%]  ***
    means: 173.219 (control), 158.576 (test)
* ssr_pe_cold_ms_jsc_nojit
    % change: -12.49% [-16.45%,  -8.53%]  ***
    means: 210.597 (control), 184.345 (test)
* ssr_pe_cold_ms_node
    % change:  -7.61% [-10.66%,  -4.56%]  ***
    means: 219.05 (control), 202.417 (test)
```

Reviewers:
@jimfb @vjeux

keyanzhang commented on src/isomorphic/classic/element/ReactElement.js:
```diff
@@ -194,12 +186,28 @@ ReactElement.createElement = function(type, config, children) {
   if (childrenLength === 1) {
-    props.children = children;
+    if (Array.isArray(children)) {
+      let childArray = children.slice(0);
```
I think it's important that the behavior stays the same across different envs so I call `slice` here regardless of the environment (prod vs. dev). Without this we won't be able to catch cases like `var children = [...]; <Comp children={children}>` since we don't want to freeze the original array reference in userland.

jimfb replied:
cc @sebmarkbage performance?

gaearon commented:
Is this incremental improvement? If so, can I merge #7410 until this gets reviewed?

jimfb commented on src/renderers/shared/stack/reconciler/__tests__/ReactComponent-test.js:
```diff
@@ -49,30 +45,39 @@ describe('ReactComponent', function() {
+
+    expect(() => {
+      ReactTestUtils.renderIntoDocument(<Wrapper>{children}</Wrapper>);
+      jest.runAllTimers();
```
Why do we need to run all timers here?

jimfb replied:
Oh, because of the new setTimeout.  Nevermind.

jimfb commented on src/isomorphic/classic/element/ReactElement.js:
```diff
@@ -194,12 +186,28 @@ ReactElement.createElement = function(type, config, children) {
+    }
   } else if (childrenLength > 1) {
-    var childArray = Array(childrenLength);
+    let childArray = Array(childrenLength);
```
This can be const, right?  The reference identity remains the same.

keyanzhang replied:
Oh yeah, good catch!

jimfb commented:
At a high level, this looks fine to me.  I flagged the slice as a performance note for Sebastian.  It would suck if we regress the prod build performance because we were trying to improve dev build performance.  If Sebastian signs off on the extra allocation/copy, I have no objections to this diff ( 👍 ).

Montana commented:
So we can't merge #7410?

gaearon commented:
I already merged #7410 in the meantime but this might be a better solution.

gaearon commented on src/renderers/shared/stack/reconciler/__tests__/ReactComponent-test.js:
```diff
@@ -49,30 +45,39 @@ describe('ReactComponent', function() {
+
+    class Wrapper extends React.Component {
+      componentDidMount() {
+        setTimeout(() => {
```
Why is `setTimeout` here necessary?

keyanzhang commented:
Rebased and updated. 
Here's a new benchmark that compares with the current master:
```
Comparing master.prod.txt (control) vs freeze.prod.txt (test)
Significant differences marked by ***
% change from control to test, with 99% CIs:

* factory_ms_jsc_jit
    % change:  -0.43% [ -5.25%,  +4.39%]
    means: 20.1321 (control), 20.0488 (test)
* factory_ms_jsc_nojit
    % change:  +0.72% [ -4.64%,  +6.08%]
    means: 18.2618 (control), 18.3966 (test)
* factory_ms_node
    % change:  +0.14% [ -2.57%,  +2.85%]
    means: 63.5866 (control), 63.6804 (test)
* ssr_pe_cold_ms_jsc_jit
    % change:  +1.98% [ -1.90%,  +5.87%]
    means: 29.2908 (control), 29.8738 (test)
* ssr_pe_cold_ms_jsc_nojit
    % change:  -1.68% [ -6.00%,  +2.64%]
    means: 31.9364 (control), 31.4052 (test)
* ssr_pe_cold_ms_node
    % change:  -0.21% [ -2.61%,  +2.20%]
    means: 58.2109 (control), 58.0934 (test)
```

```
Comparing master.dev.txt (control) vs freeze.dev.txt (test)
Significant differences marked by ***
% change from control to test, with 99% CIs:

* factory_ms_jsc_jit
    % change:  -2.74% [ -7.44%,  +1.97%]
    means: 27.3811 (control), 26.6391 (test)
* factory_ms_jsc_nojit
    % change:  +0.33% [ -4.34%,  +5.00%]
    means: 24.6727 (control), 24.7591 (test)
* factory_ms_node
    % change:  -0.37% [ -2.90%,  +2.17%]
    means: 90.5355 (control), 90.2092 (test)
* ssr_pe_cold_ms_jsc_jit
    % change:  -4.46% [ -7.64%,  -1.27%]  ***
    means: 87.2714 (control), 83.392 (test)
* ssr_pe_cold_ms_jsc_nojit
    % change:  -5.38% [ -9.78%,  -0.98%]  ***
    means: 93.9902 (control), 88.9515 (test)
* ssr_pe_cold_ms_node
    % change:  -4.82% [ -6.69%,  -2.95%]  ***
    means: 132.504 (control), 126.122 (test)
```

keyanzhang commented:
Friendly ping 😀: I just rebased and updated this PR. Do we want to take this approach to fix #7424?

To sum up, it calls `.slice()` when people explicitly pass an array as the `children` prop (`<Comp children={[...]}/>` or `<Comp>{[...]}<Comp/>`) and freezes the sliced copy, so that we won't accidentally freeze userland data. We call `slice` in both prod and dev to align their behaviors, which regresses the prod build perf a little bit. A benchmark can be found above.

gaearon commented:
Can you please help me understand how to read this profiler script output?
I don’t know if `+` is good or `-` is good.

gaearon commented on src/isomorphic/classic/element/ReactElement.js:
```diff
@@ -231,13 +223,30 @@ ReactElement.createElement = function(type, config, children) {
   if (childrenLength === 1) {
-    props.children = children;
+    if (Array.isArray(children)) {
+      childArray = children.slice(0);
```
Does this mean that if we pass them through multiple levels of components, the same array could get `slice(0)`d many times?

keyanzhang replied:
Yes, that's unfortunate. :( Do you think there's a better way to do this? 

However, it should be more performant than the current approach, which slices in the `ReactElement()` function (https://github.com/facebook/react/blob/master/src/isomorphic/classic/element/ReactElement.js#L138) for every component that has more than one child. The new one slices in `createElement` and only when an user explicitly passes an array as `children`. If this makes sense then I guess the question is whether we want to slice in prod.

gaearon replied:
I’m mostly worried about prod here.
I don’t think we can afford to do more allocations in prod.

keyanzhang commented:
Yeah, sorry, `-` is good and `+` is bad. cc @spicyj

gaearon commented:
What does the benchmark measure?
Does it include memory usage?

keyanzhang commented:
No, it only measures time. The percentages (`-2.74% [ -7.44%,  +1.97%]`) are mean, low, and high (https://github.com/facebook/react/blob/master/scripts/bench/analyze.py#L102-L107).

gaearon commented:
I don’t think allocating more in production is the way to go here.
I think we should either keep this warning as is, or write a lint rule that warns on

```js
this.props.children.<mutatingMethod>()
```

or

```js
(var|const|let) <name> = (this.)?props.children;
<name>.{mutatingMethod}()
```

If this covers most cases and gets included into the default set of `eslint-plugin-react`, Create React App, and www codebase, we can remove the warning altogether IMO.

spicyj commented:
It probably suffices to freeze the array when we create it and do nothing otherwise, don't you think?

gaearon commented:
I’m not sure when people would try to `push` things into `children` so hard to say.
Maybe @jimfb can show a few more examples of how people violated this warning in practice.
I thought that pushing things into an array created by `map()` is one of those use cases.

gaearon commented:
But I’m OK with interpretation above. We don’t have to catch every possible mistake, and what @spicyj suggested seems like good compromise to me.

keyanzhang commented:
Sounds good to me! I agree that we shouldn't introduce an overhead in prod.

gaearon commented:
Could you rebase pls?

keyanzhang commented:
Sure, just rebased.

gaearon commented:
Thanks!

zpao commented:
FYI: There's *no way* this is a patch level change - it changes what was a warning to start throwing. Sure, it's dev-only but that's not ok to take in a patch release (we avoid introducing new warnings in patch releases, we definitely can't start throwing).

Let's be a bit more conservative with the semver assignment so we don't sneak breakages in at the wrong time.

just-boris commented:
@zpao does that mean that there will be `15.4` release soon?

gaearon commented:
I'll be more conservative, thanks for correcting!
